airflow/models/taskinstancehistory.py

```diff
@@ -0,0 +1,130 @@
+    run_id = Column(StringID(), nullable=False)
+    map_index = Column(Integer, nullable=False, server_default=text("-1"))
+    try_number = Column(Integer, nullable=False)
+    # The rest of the columns are kept in sync with TaskInstance, added below
+
+    def __init__(
+        self,
+        ti: TaskInstance | TaskInstancePydantic,
+        state: str | None = None,
+    ):
+        super().__init__()
+        for column in self.__table__.columns:
+            if column.name == "id":
+                continue
+            setattr(self, column.name, getattr(ti, column.name))
+
+        if state:
+            self.state = state
+
+    __table_args__ = (
+        ForeignKeyConstraint(
+            [dag_id, task_id, run_id, map_index],
+            [
+                "task_instance.dag_id",
+                "task_instance.task_id",
+                "task_instance.run_id",
+                "task_instance.map_index",
+            ],
+            name="task_instance_history_ti_fkey",
+            ondelete="CASCADE",
+        ),
+        UniqueConstraint(
+            "dag_id",
+            "task_id",
+            "run_id",
+            "map_index",
+            "try_number",
+            name="task_instance_history_dtrt_uq",
+        ),
+    )
+
+    @staticmethod
+    @provide_session
+    def record_ti(ti: TaskInstance, session: NEW_SESSION = None) -> None:
+        """Record a TaskInstance to TaskInstanceHistory."""
+        exists_q = session.scalar(
+            select(func.count(TaskInstanceHistory.task_id)).where(
+                TaskInstanceHistory.dag_id == ti.dag_id,
+                TaskInstanceHistory.task_id == ti.task_id,
+                TaskInstanceHistory.run_id == ti.run_id,
+                TaskInstanceHistory.map_index == ti.map_index,
+                TaskInstanceHistory.try_number == ti.try_number,
+            )
+        )
+        if exists_q:
+            return
+        ti_history_state = ti.state
+        if ti.state not in State.finished:
+            ti_history_state = TaskInstanceState.FAILED
+            ti.end_date = timezone.utcnow()
+            ti.set_duration()
+        ti_history = TaskInstanceHistory(ti, state=ti_history_state)
+        session.add(ti_history)
+
+
+def _copy_column(column):
+    return Column(
+        column.type,
+        nullable=column.nullable,
+        default=column.default,
+        autoincrement=column.autoincrement,
+        unique=column.unique,
+        index=column.index,
+        primary_key=None,
+        server_default=column.server_default,
+        server_onupdate=column.server_onupdate,
+        doc=column.doc,
+        comment=column.comment,
+        info=column.info,
+    )
+
+
+# Add remaining columns from TaskInstance to TaskInstanceHistory, as we want to keep them in sync
+for column in TaskInstance.__table__.columns:
+    if column.name not in TaskInstanceHistory.__table__.columns:
+        setattr(TaskInstanceHistory, column.name, _copy_column(column))
```

Comment:
I wonder if it is better to actually duplicate the fields instead, this plus SQLAlchemy already having a ton of magic makes me a bit nervous. Maybe it is better to actually duplicate the fields literally and keep things in sync with a pre-commit hook?

Reply:
What do you think about this @jedcunningham

Reply:
I'm definitely open to switching this up, but I'd like to do it in a follow up PR instead.